app/controllers/shopify_app/callback_controller.rb

```diff
@@ -6,38 +6,57 @@ class CallbackController < ActionController::Base
     include ShopifyApp::LoginProtection
 
     def callback
-      unless auth_hash
-        return respond_with_error
-      end
+      return respond_with_error if invalid_request?
+
+      store_access_token_and_build_session
 
-      if jwt_request? && !valid_jwt_auth?
-        Rails.logger.debug("[ShopifyApp::CallbackController] Invalid JWT auth detected.")
-        return respond_with_error
+      if start_user_token_flow?
+        return respond_with_user_token_flow
       end
 
+      perform_post_authenticate_jobs
+
+      respond_successfully
+    end
+
+    private
+
+    def respond_successfully
       if jwt_request?
-        Rails.logger.debug("[ShopifyApp::CallbackController] JWT request detected. Setting shopify session...")
-        set_shopify_session
         head(:ok)
       else
-        Rails.logger.debug("[ShopifyApp::CallbackController] Not a JWT request. Resetting session options...")
-        reset_session_options
-        set_shopify_session
+        redirect_to(return_address)
+      end
+    end
+
+    def respond_with_user_token_flow
+      redirect_to(login_url_with_optional_shop)
+    end
 
-        if redirect_for_user_token?
-          Rails.logger.debug("[ShopifyApp::CallbackController] Redirecting for user token...")
-          return redirect_to(login_url_with_optional_shop)
-        end
+    def prevent_session_fixation
+      reset_session_options if native_browser_request?
+    end
 
-        install_webhooks
-        install_scripttags
-        perform_after_authenticate_job
+    def store_access_token_and_build_session
+      prevent_session_fixation if native_browser_request?
```

Comment:
Do we need to check `native_browser_request?` here, if it's also checked in `prevent_session_fixation`?

Reply:
You're right! Just refactored it.